goldmane/pkg/aggregator/index.go

```diff
@@ -60,42 +64,81 @@ type IndexFindOpts struct {
 	filter *proto.Filter
 }
 
-func (idx *index[E]) List(opts IndexFindOpts) []*types.Flow {
+func (idx *index[E]) List(opts IndexFindOpts) ([]*types.Flow, int) {
 	logrus.WithFields(logrus.Fields{
 		"opts": opts,
 	}).Debug("Listing flows from index")
 
 	var matchedFlows []*types.Flow
+	var totalMatchedCount int
 
-	// Find the index into the DiachronicFlow array from which to start the search based on the provided page number and limit.
-	// We need to iterate through the DiachronicFlows to find the starting point, since not every DiachronicFlow is necessarily a match
-	// for the given options.
-	var i int
-	if opts.page > 0 {
-		target := opts.page * opts.limit
-		for j := 0; j < len(idx.diachronics); j++ {
-			if idx.evaluate(idx.diachronics[j], opts) != nil {
-				i++
-			}
-			if int64(i) == target {
-				break
+	target := int(opts.page * opts.pageSize)
+
+	// Iterate through the DiachronicFlows and evaluate each one until we reach the limit or the end of the list.
+	for _, diachronic := range idx.diachronics {
+		flow := idx.evaluate(diachronic, opts)
+		if flow != nil {
+			// increment the count regardless of whether we're including the key, as we need a total matching count.
+			totalMatchedCount++
+
+			// The target represents the index of the last key on the previous page, so we need to ensure that we're above
+			// that index before we start collecting flows.
+			if totalMatchedCount > target {
+				// Only append the key if
+				if opts.pageSize == 0 || int64(len(matchedFlows)) < opts.pageSize {
+					matchedFlows = append(matchedFlows, flow)
+				}
 			}
 		}
 	}
 
+	return matchedFlows, calculatePageCount(totalMatchedCount, int(opts.pageSize))
+}
+
+// Keys retrieves a unique set of keys matching the given index options.
+func (idx *index[E]) Keys(opts IndexFindOpts) ([]E, int) {
```

Comment:
I think Keys is a confusing name for this method of the Index and has thrown me off a few times reading through the code now - a `Key` is a unique identifier for an object, which isn't what this function returns. Plus we already use Key pretty extensively to refer to the FlowKey, which is an identifier for a distinct Flow object. 

Rather, this function returns a deduplicated set of fundamentally non-unique values currently within the index.

Maybe `ValueSet`, or `SortValueSet`?

This is tied with my other comment about tightening up the terminology here for things like `keyFunc`, etc.